### ISSUE TITLE
Multiple Ganga sessions

### ISSUE THREAD

**Issue by rob-c** [ganga-devs/ganga#236]
Hi all,

As per #235 I think it's worth bringing up multiple concurrent Ganga sessions.

I would like to propose that if Ganga finds an active session lock that it should by default disable the monitoring thread as well as informing the user. This should prevent some level of repository clashes that we see if someone forgets and launches 2 versions of ganga in parallel.

I think a short message informing the user that they've started in a _read-only_ mode and that they can enable monitoring would be useful if we do this.

I think this also strongly merits that we check that both ```enableMonitoring()``` and ```disableMonitoring()``` do what is advertised.
I've been working on a test for this and have spotted a few minor problems which I'd like to put into a smaller PR in the near future.

Does everyone else think this sounds sane?

Thanks,

Rob

**Comment by alexanderrichards** [comment]
sounds very sensible to me. We already tell everyone if they are starting a extra session to disable the monitoring, it makes sense for us to do it for them.

We will need to make sure that we don't end up in (or at least correctly handle) the situation where there is a rogue left over lock even though no sessions are active. I guess this is hand in hand with what you mention:

> I think this also strongly merits that we check that both enableMonitoring() and disableMonitoring() do what is advertised.

such that we can correctly re-enable the monitoring if needed.

anyway I'm in favour of this.

**Comment by milliams** [comment]
I agree in principle but I'd argue that starting with monitoring disabled is not the same as starting in read-only mode. Without monitoring you can still add and remove jobs etc.

I would personally take a stronger stance and refuse to start if there is another instance running and give the user the information they need to check whether it is true (i.e. the PID of the other instance, perhaps which lxplus node it is running on etc.) That way they can check if it really is running, close it if it is or remove the lock file if it is not.

**Comment by rob-c** [comment]
I said _read-only_ not `read-only` as it's a matter of perspective, I was trying to make it clear what's happening.

I think we need to address the issue of stale lock files separately.
Currently the timeout for a lockfile is set a little high. But for good reasons (which may or may not be relevant anymore).
I would maybe suggest advertising a command to force checking of the lockfiles from an interactive IPython session but this is a little invasive.

__No__, I would not want ganga to not to start due to stale lockfiles. (or more explicitly ganga, should _always_ get to an IPython prompt unless a critical error occurs).
Ganga will then refuse to start if users get a broken connection to lxplus and then we have _insanity_ from the users perspective.
There are many people who just ignore this warning/error (for better or worse).

**Comment by alexanderrichards** [comment]
indeed read-only may have been a bit misleading, it would essentially be a more formal way of running the current (not-really) recommended way of having multiple sessions. Basically no monitoring support. The warning would probably need to be along these lines.

Whether or not we want to make this a stronger "No multiple sessions" I don't really mind but this suggestion would be good as a minimum way of addressing potential sources of bugs due to colliding monitoring loops.

**Comment by drmarkwslater** [comment]
I think there's an argument to be made for NOT starting Ganga if there are multiple (active) sessions detected, but then there's also an argument for starting Ganga but with monitoring disabled (which I actually thought was what happened at the moment but I guess I'm wrong). Maybe have a config option so users who want concurrent sessions can have them but those that don't have some protection in place?

**Comment by rob-c** [comment]
The advice for a long time has been that ```ganga``` on one machine and ```ganga --no-mon``` on another machine is relatively safe for lxplus users (read LHCb).
I'd rather not suddenly change this as there would likely be strong backlash from some quarters.

**Comment by rob-c** [comment]
It would be nice to have method to turn off the monitoring if we detect a concurrent Ganga session before the 6.1.19 release. This shouldn't be too difficult and can be probably tied into the construction of the Registries, do people think this is best to get in 6.1.19 or should we hold off with this until a future release? (I don't mind either way)

**Comment by rob-c** [comment]
This can probably be bumped until 6.1.20 All I'm aiming to do with this is make sure the monitoring loop is deactivated when multiple other sessions are detected but this feels like a nicety at this point.

**Comment by afortiorama** [comment]
It is an enhancement.

**Comment by rob-c** [comment]
This ties in with #235 so I'll mention it here so we know to close it when address this.

**Comment by rob-c** [comment]
@afortiorama I agree it is an enhancement but we're keen to get the next release out it's been a while now and there are a lot of other things which it addresses

**Comment by rob-c** [comment]
I'd like to see this get in in 6.1.21 as I bump this. Hopefully the dev work will be at a slightly more relaxed pace then. I'll give the issue a milestone again once it's actually got a PR for it.
